service/src/main/java/com/theokanning/openai/service/ResponseBodyCallback.java

```diff
@@ -40,61 +40,79 @@ public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response)
 
         try {
             if (!response.isSuccessful()) {
-                HttpException e = new HttpException(response);
-                ResponseBody errorBody = response.errorBody();
-
-                if (errorBody == null) {
-                    throw e;
-                } else {
-                    OpenAiError error = mapper.readValue(
-                            errorBody.string(),
-                            OpenAiError.class
-                    );
-                    throw new OpenAiHttpException(error, e, e.code());
-                }
+                handleUnsuccessfulResponse(response);
+                return;
             }
 
             InputStream in = response.body().byteStream();
             reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
-            String line;
-            SSE sse = null;
+            parseSSE(reader);
+
+            emitter.onComplete();
+
+        } catch (Throwable t) {
+            onFailure(call, t);
+        } finally {
+            if (reader != null) {
+                try {
+                    reader.close();
+                } catch (IOException e) {
+                    // do nothing
+                }
+            }
+        }
+    }
 
+    private void handleUnsuccessfulResponse(Response<ResponseBody> response) throws IOException {
+        HttpException e = new HttpException(response);
+        ResponseBody errorBody = response.errorBody();
+
+        if (errorBody == null) {
+            throw e;
+        } else {
+            OpenAiError error = mapper.readValue(
+                    errorBody.string(),
+                    OpenAiError.class
+            );
+            throw new OpenAiHttpException(error, e, e.code());
+        }
+    }
```

Comment:
Nit: It can be optimized  

```java

try {
    ResponseBody errorBody = response.errorBody();
    if (errorBody != null) {
        OpenAiError error = mapper.readValue(errorBody.string(), OpenAiError.class);
        throw new OpenAiHttpException(error, new HttpException(response), response.code());
    }
} catch (IOException ex) {
    throw new HttpException(response);
}
```